### PR TITLE
fix(proxy): resolving alias domain name for multiple deployment

### LINKED ISSUE #210: Multiple deployments: can not resolve "alias" domain name
Screen_Shot_2021-09-25_at_5_33_07_AM

Assuming that the data in the tf-next-proxy-config table is as above, if the user uses the `v6.mydomain.com` domain, the `aliasedTo`(`cktx6iq3i0000j9pa0k6n3ona`) value must be used as the `deploymentIdentifier` to find the object in s3.

https://github.com/milliHQ/terraform-aws-next-js/blob/289963f3d76f25ee5c36b4662225923032bae1f2/packages/proxy/src/handler.ts#L231

But there is no part that refers to `aliasedTo`, so this aliase domain name doesn't work.

### PR DESCRIPTION
**Pull request by b6pzeusbc54tvhw5jgpyw8pwz2x6gs**
resolve #210 
There was a little big change from `proxyConfig` to `deployment`. Any inquiries or reviews are welcome.

- Defined new types `Deployment`, `DeploymentInDB`.
- Changed `fetchProxyConfig` to `fetchDeployment` to refer `aliasedTo`.
- To use simple parameters when getting deployment information,
  Changed to use `new DynamoDB.DocumentClient()` instead of `new DynamoDB()`.
      - https://stackoverflow.com/a/57807642/5568628
- Cached ddbClient object per region.
- Added `keepAlive: true` agent setting into dynamoDB Client for performance. However, #197 still needs to be resolved.
(@endor @ofhouse )

### CODE REVIEW

File: packages/proxy/src/handler.ts [hunk 1]
```diff
@@ -106,12 +106,15 @@ export async function handler(event: CloudFrontRequestEvent) {
...
-      proxyConfig = await fetchProxyConfig(
+      deployment = await fetchDeployment(
         configEndpointURL.toString(),
         configTable,
         configRegion,
         deploymentIdentifier,
       );
+      if (deployment && deployment.aliasedTo) {
+        deploymentIdentifier = deployment.aliasedTo;
+      }
```
Comment (b6pzeusbc54tvhw5jgpyw8pwz2x6gs) [inline packages/proxy/src/handler.ts:117]: 8n this part, if there is an aliasedTo value, it is assigned to the `deploymentIdentifier`.

File: packages/proxy/src/util/fetch-deployment.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        // If something fails, we continue with the default
+        // behaviour of trying to fetch the config from the bucket
+      }
+    }
+  }
+
+  const res = await fetchTimeout(FETCH_TIMEOUT, endpointUrl);
+  const proxyConfig = (await res.json()) as ProxyConfig;
+  const deployment: Deployment = { alias: '-', proxyConfig };
+  return deployment;
```
Comment (b6pzeusbc54tvhw5jgpyw8pwz2x6gs) [inline packages/proxy/src/util/fetch-deployment.ts:41]: Will this part be deprecated later?

File: packages/proxy/src/handler.ts (hunk 1 above)
Comment (endor) [reply packages/proxy/src/handler.ts:117]: Why do we need this?

File: packages/proxy/src/util/fetch-deployment.ts (hunk 2 above)
Comment (endor) [reply packages/proxy/src/util/fetch-deployment.ts:41]: The idea is to deprecate it and have multiple deployments as the only mode of operation.

File: packages/proxy/src/handler.ts (hunk 1 above)
Comment (b6pzeusbc54tvhw5jgpyw8pwz2x6gs) [reply packages/proxy/src/handler.ts:117]: Currently there is issue #210 . In the **complete** example, reloading the browser on the `about` page reproduces the problem that the object in s3 cannot be found.

File: packages/proxy/src/util/fetch-deployment.ts (hunk 2 above)
Comment (b6pzeusbc54tvhw5jgpyw8pwz2x6gs) [reply packages/proxy/src/util/fetch-deployment.ts:41]: Good 👍

File: packages/proxy/src/handler.ts (hunk 1 above)
Comment (b6pzeusbc54tvhw5jgpyw8pwz2x6gs) [reply packages/proxy/src/handler.ts:117]: @endor @ofhouse As far as I know, without this patch, there is no way to get the static file in s3 using alias domain. please review this PR.

**Comment by b6pzeusbc54tvhw5jgpyw8pwz2x6gs** [comment]
@ofhouse Thanks for sharing the background. I understand the current situation well. 

After setting up the infrastructure with Terraform, I think it is very important to enable subsequent deployments without Terraform. It's probably not as good as @endor does, but if I have the time, can I work on the **Multiple Deployment** issues one by one in the v1.x branch?

**Comment by ofhouse** [comment]
@b6pzeusbc54tvhw5jgpyw8pwz2x6gs Yes, for sure!
If you have time to work on these issues you are welcome to contribute 😊

**Pull request merged**